### ISSUE TITLE
Add --coverage-only-library, make default coverage behavior simpler

### ISSUE THREAD

**Issue by mgsloan** [commercialhaskell/stack#1191]
I'm revisiting https://github.com/commercialhaskell/stack/issues/634 ("stack test --coverage shouldn't include tests in reports")

* It's causing issues on 7.10 - https://github.com/commercialhaskell/stack/issues/1181

* It's common for packages to define their tests in such a way that they end up building the library directly with the test.

In light of this, I think it's a good idea to put this behavior behind a flag like `--coverage-only-library`.

**Comment by mgsloan** [comment]
Pinging @mboes, does this seem reasonable?

**Comment by mboes** [comment]
Fair enough. But I assume that, if enabled, one is likely to run into #1181?

**Comment by mgsloan** [comment]
I've decided not to do this.  This stuff got a bit complicated, I don't expect this to be 100% clear:

I've reinstated my old solution to #1181 with a fix for #1162, so actually I think things should be fine now.

I tried out implementing this, but ran into issues with cabal files that have multiple executables, due to https://ghc.haskell.org/trac/ghc/ticket/1853#comment:23 .  Since I can't use the hpc program's `--include` argument to leave out the unqualified modules when there isn't a library stanza (#1162), my solution for now is to munge all tix files to remove unqualified modules.

Since I'm removing all unqualified modules, we don't really get the benefit of having a solution for packages that build their library code directly into their tests.
